subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/UpgradePropertyInstrumentationExtension.java

```diff
@@ -0,0 +1,250 @@
+import org.gradle.internal.instrumentation.model.CallableInfoImpl;
+import org.gradle.internal.instrumentation.model.CallableKindInfo;
+import org.gradle.internal.instrumentation.model.ImplementationInfo;
+import org.gradle.internal.instrumentation.model.ImplementationInfoImpl;
+import org.gradle.internal.instrumentation.model.ParameterInfo;
+import org.gradle.internal.instrumentation.model.ParameterInfoImpl;
+import org.gradle.internal.instrumentation.model.RequestExtra;
+import org.gradle.internal.instrumentation.processor.codegen.InstrumentationCodeGenerator;
+import org.gradle.internal.instrumentation.processor.codegen.RequestGroupingInstrumentationClassGenerator;
+import org.gradle.internal.instrumentation.processor.extensibility.AnnotatedMethodReaderExtension;
+import org.gradle.internal.instrumentation.processor.extensibility.ClassLevelAnnotationsContributor;
+import org.gradle.internal.instrumentation.processor.extensibility.CodeGeneratorContributor;
+import org.gradle.internal.instrumentation.processor.modelreader.api.CallInterceptionRequestReader.Result.InvalidRequest;
+import org.gradle.internal.instrumentation.processor.modelreader.api.CallInterceptionRequestReader.Result.Success;
+import org.gradle.internal.instrumentation.processor.modelreader.impl.AnnotationUtils;
+import org.jetbrains.annotations.NotNull;
+import org.objectweb.asm.Type;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.type.TypeMirror;
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.gradle.internal.instrumentation.model.ParameterKindInfo.METHOD_PARAMETER;
+import static org.gradle.internal.instrumentation.processor.codegen.TypeUtils.*;
+import static org.gradle.internal.instrumentation.processor.modelreader.impl.TypeUtils.extractType;
+
+public class UpgradePropertyInstrumentationExtension
+    extends RequestGroupingInstrumentationClassGenerator
+    implements AnnotatedMethodReaderExtension, ClassLevelAnnotationsContributor, CodeGeneratorContributor {
+
+    private static final Type PROPERTY = Type.getType(Property.class);
+    private static final Type REGULAR_FILE_PROPERTY = Type.getType(RegularFileProperty.class);
+    private static final Type DIRECTORY_PROPERTY = Type.getType(DirectoryProperty.class);
+    private static final Type DEFAULT_TYPE = Type.getType(UpgradedProperty.SameAsGenericType.class);
+    private static final String INTERCEPTOR_DECLARATION_CLASS_NAME = "org.gradle.internal.classpath.InterceptorDeclaration_JvmBytecodeImplCodeQuality";
```

Comment:
UPD: I've just found a TODO in a later commit. Feel free to resolve this.

Reply:
Well, it is an API of some sort, and I'd say we should make the sharing of the class names as visible as we reasonably can, because implicit connections will likely be confusing for someone making changes in the subsystem for the first time.

Reply:
Move to instrumentation-api